src/crypto/crypto_util.h

```diff
@@ -0,0 +1,680 @@
+// contents are zeroed.
+class ByteSource {
+ public:
+  ByteSource() = default;
+  ByteSource(ByteSource&& other) noexcept;
+  ~ByteSource();
+
+  ByteSource& operator=(ByteSource&& other) noexcept;
+
+  const char* get() const;
+
+  template <typename T>
+  const T* data() const { return reinterpret_cast<const T*>(get()); }
+
+  size_t size() const;
+
+  inline operator bool() const {
+    return data_ != nullptr;
+  }
+
+  inline BignumPointer ToBN() const {
+    return BignumPointer(BN_bin2bn(
+        reinterpret_cast<const unsigned char*>(get()),
+        size(),
+        nullptr));
+  }
+
+  // Creates a v8::BackingStore that takes over responsibility for
+  // any allocated data. The ByteSource will be reset with size = 0
+  // after being called.
+  std::unique_ptr<v8::BackingStore> ReleaseToBackingStore();
+
+  v8::Local<v8::ArrayBuffer> ToArrayBuffer(Environment* env);
+
+  void reset();
+
+  // Allows an Allocated ByteSource to be truncated.
+  inline void Resize(size_t newsize) {
+    CHECK_LE(newsize, size_);
+    CHECK_NOT_NULL(allocated_data_);
+    char* new_data_ = ReallocOpenSSL<char>(allocated_data_, newsize);
+    data_ = allocated_data_ = new_data_;
+    size_ = newsize;
+  }
+
+  static ByteSource Allocated(char* data, size_t size);
+  static ByteSource Foreign(const char* data, size_t size);
+
+  static ByteSource FromEncodedString(Environment* env,
+                                      v8::Local<v8::String> value,
+                                      enum encoding enc = BASE64);
+
+  static ByteSource FromStringOrBuffer(Environment* env,
+                                       v8::Local<v8::Value> value);
+
+  static ByteSource FromString(Environment* env,
+                               v8::Local<v8::String> str,
+                               bool ntc = false);
+
+  static ByteSource FromBuffer(v8::Local<v8::Value> buffer,
+                               bool ntc = false);
+
+  static ByteSource FromBIO(const BIOPointer& bio);
+
+  static ByteSource NullTerminatedCopy(Environment* env,
+                                       v8::Local<v8::Value> value);
+
+  static ByteSource FromSymmetricKeyObjectHandle(v8::Local<v8::Value> handle);
+
+  ByteSource(const ByteSource&) = delete;
+  ByteSource& operator=(const ByteSource&) = delete;
+
+  static ByteSource FromSecretKeyBytes(
+      Environment* env, v8::Local<v8::Value> value);
+
+ private:
+  const char* data_ = nullptr;
+  char* allocated_data_ = nullptr;
+  size_t size_ = 0;
+
+  ByteSource(const char* data, char* allocated_data, size_t size);
+};
+
+enum CryptoJobMode {
+  kCryptoJobAsync,
+  kCryptoJobSync
+};
+
+inline CryptoJobMode GetCryptoJobMode(v8::Local<v8::Value> args) {
+  CHECK(args->IsUint32());
+  uint32_t mode = args.As<v8::Uint32>()->Value();
+  CHECK_LE(mode, kCryptoJobSync);
+  return static_cast<CryptoJobMode>(mode);
+}
+
+template <typename CryptoJobTraits>
+class CryptoJob : public AsyncWrap, public ThreadPoolWork {
```

Comment:
Definitely a fan of making this a propert `AsyncWrap` :)

You can remove `AsyncWrapObject` from `src/async_wrap.cc` now, if you like, I don’t think it’s generally useful and this approach here is the better one imo :)

Reply:
Yeah, I'll do that in a separate commit once this lands